File: on_audio_query/android/src/main/kotlin/com/lucasjosino/on_audio_query/methods/queries/SongsQuery.kt
package com.lucasjosino.on_audio_query.methods.queries

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lucasjosino.on_audio_query.controllers.PermissionController
import com.lucasjosino.on_audio_query.methods.helper.QueryHelper
import com.lucasjosino.on_audio_query.types.checkAudioType
import com.lucasjosino.on_audio_query.types.checkAudiosUriType
import com.lucasjosino.on_audio_query.types.sorttypes.checkSongSortType
import com.lucasjosino.on_audio_query.utils.songProjection
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** SongsQuery */
class SongsQuery : ViewModel() {

    // Main parameters
    private val helper = QueryHelper()
    private var selection: String = ""

    // None of this methods can be null.
    private lateinit var uri: Uri
    private lateinit var resolver: ContentResolver
    private lateinit var sortType: String

    @Suppress("UNCHECKED_CAST")
    fun init(
        context: Context,
        //
        result: MethodChannel.Result? = null,
        call: MethodCall? = null,
        //
        sink: EventChannel.EventSink? = null,
        args: Map<*, *>? = null
    ) {
        resolver = context.contentResolver

        val pSortType: Int?
        val pOrderType: Int
        val pIgnoreCase: Boolean
        val pUri: Int

        val toQuery: MutableMap<Int, ArrayList<String>>
        val toRemove: MutableMap<Int, ArrayList<String>>
        val type: MutableMap<Int, Int>

        if (sink != null && args != null) {
            pSortType = args["sortType"] as Int?
            pOrderType = args["orderType"] as Int
            pIgnoreCase = args["ignoreCase"] as Boolean
            pUri = args["uri"] as Int

            toQuery = args["toQuery"] as MutableMap<Int, ArrayList<String>>
            toRemove = args["toRemove"] as MutableMap<Int, ArrayList<String>>
            type = args["type"] as MutableMap<Int, Int>
        } else {
            pSortType = call!!.argument<Int>("sortType")
            pOrderType = call.argument<Int>("orderType")!!
            pIgnoreCase = call.argument<Boolean>("ignoreCase")!!
            pUri = call.argument<Int>("uri")!!

            toQuery = call.argument<MutableMap<Int, ArrayList<String>>>("toQuery")!!
            toRemove = call.argument<MutableMap<Int, ArrayList<String>>>("toRemove")!!
            type = call.argument<MutableMap<Int, Int>>("type")!!
        }

        // Sort: Type and Order.
        sortType = checkSongSortType(
            pSortType,
            pOrderType,
            pIgnoreCase
        )

        // Check uri:
        //   * [0]: External.
        //   * [1]: Internal.
        uri = checkAudiosUriType(pUri)

        // Add item/items to 'query'.
        for ((id, values) in toQuery) {
            for (value in values) {
                selection += songProjection[id] + " LIKE '%" + value + "%' " + "AND "
            }
        }

        // Remove item/items from 'query'.
        for ((id, values) in toRemove) {
            for (value in values) {
                selection += songProjection[id] + " NOT LIKE '%" + value + "%' " + "AND "
            }
        }

        // Add/Remove audio type. E.g: Is Music, Notification, Alarm, etc..
        for (audioType in type) {
            selection += checkAudioType(audioType.key) + "=" + "${audioType.value} " + "AND "
        }

        // Remove the 'AND ' keyword from selection.
        if (selection.endsWith("AND ")) {
            selection = selection.removeSuffix("AND ")
        }

        // Init the 'query'.
        querySongs(context, result, sink)
    }

    private fun querySongs(
        context: Context,
        result: MethodChannel.Result?,
        sink: EventChannel.EventSink?
    ) {
        // Request permission status from the 'main' method.
        val hasPermission: Boolean = PermissionController().permissionStatus(context)

        // We cannot 'query' without permission so, throw a PlatformException.
        // Only one 'channel' will be 'functional'. If is null, ignore, if not, send the error.
        if (!hasPermission) {
            // Method from 'EventChannel' (observer)
            sink?.error(
                "403",
                "The app doesn't have permission to read files.",
                "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
            )
            // Method from 'MethodChannel' (method)
            result?.error(
                "403",
                "The app doesn't have permission to read files.",
                "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
            )

            // 'Exit' the function
            return
        }

        // Query everything in background for a better performance.
        viewModelScope.launch {
            // Start 'querying'.
            val resultSongList: ArrayList<MutableMap<String, Any?>> = loadSongs()

            // After loading the information, send the 'result'.
            sink?.success(resultSongList)
            result?.success(resultSongList)
        }
    }

    //Loading in Background
    private suspend fun loadSongs(): ArrayList<MutableMap<String, Any?>> =
        withContext(Dispatchers.IO) {

            // Setup the cursor with [uri], [projection] and [sortType].
            val cursor = resolver.query(uri, songProjection, selection, null, sortType)
            // Empty list.
            val songList: ArrayList<MutableMap<String, Any?>> = ArrayList()

            // For each item(song) inside this "cursor", take one and "format"
            // into a [Map<String, dynamic>].
            while (cursor != null && cursor.moveToNext()) {
                val tempData: MutableMap<String, Any?> = HashMap()
                for (audioMedia in cursor.columnNames) {
                    tempData[audioMedia] = helper.loadSongItem(audioMedia, cursor)
                }

                //Get a extra information from audio, e.g: extension, uri, etc..
                val tempExtraData = helper.loadSongExtraInfo(uri, tempData)
                tempData.putAll(tempExtraData)

                songList.add(tempData)
            }

            // Close cursor to avoid memory leaks.
            cursor?.close()
            // After finish the "query", go back to the "main" thread(You can only call flutter
            // inside the main thread).
            return@withContext songList
        }
}

//Extras:

// * Query only audio > 60000 ms [1 minute]
// Obs: I don't think is a good idea, some audio "Non music" have more than 1 minute
//query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, projection, MediaStore.Audio.Media.DURATION +
// ">= 60000", null, checkSongSortType(sortType!!))

// * Query audio with limit, used for better performance in tests
//MediaStore.Audio.Media.TITLE + " LIMIT 4"

// * All projection types in android [Audio]
//I/AudioCursor[All]: [
// title_key,
// instance_id,
// duration,
// is_ringtone,
// album_artist,
// orientation,
// artist,
// height,
// is_drm,
// bucket_display_name,
// is_audiobook,
// owner_package_name,
// volume_name,
// title_resource_uri,
// date_modified,
// date_expires,
// composer,
// _display_name,
// datetaken,
// mime_type,
// is_notification,
// _id,
// year,
// _data,
// _hash,
// _size,
// album,
// is_alarm,
// title,
// track,
// width,
// is_music,
// album_key,
// is_trashed,
// group_id,
// document_id,
// artist_id,
// artist_key,
// is_pending,
// date_added,
// is_podcast,
// album_id,
// primary_directory,
// secondary_directory,
// original_document_id,
// bucket_id,
// bookmark,
// relative_path
// ]

File: on_audio_query/ios/Classes/methods/queries/SongsQuery.swift
import Flutter
import MediaPlayer

class SongsQuery {
    // Main parameters
    private var args: [String: Any]
    private var result: FlutterResult?
    private var sink: FlutterEventSink?
    
    // Song projection (to filter).
    private let songProjection: [String?] = [
        "_id",
        "_data",
        "_display_name",
        nil,
        "album",
        nil,
        "album_id",
        "artist",
        "artist_id",
        nil,
        "composer",
        nil,
        nil,
        nil,
        "title",
        nil,
        nil,
        nil,
        nil,
        nil,
        nil,
        nil,
    ]
    
    init(
        // Call from 'MethodChannel' (method).
        call: FlutterMethodCall? = nil,
        result: FlutterResult? = nil,
        // Call from 'EventChannel' (observer).
        sink: FlutterEventSink? = nil,
        args: [String: Any]? = nil
    ) {
        // Get all arguments inside the map.
        self.args = sink != nil ? args! : (call!.arguments as! [String: Any])
        self.sink = sink
        self.result = result
    }
    
    func querySongs() {
        // The sortType. If 'nil', will be set as [Title].
        let sortType = args["sortType"] as? Int ?? 0
        
        // Choose the type(To match android side, let's call "cursor").
        let cursor = MPMediaQuery.songs()
        
        // Make sure that only audios with 'music' type will be 'queried'.
        cursor.addFilterPredicate(MPMediaPropertyPredicate.init(
            value: MPMediaType.music.rawValue,
            forProperty: MPMediaItemPropertyMediaType,
            comparisonType: .equalTo
        ))
        
        // This filter will avoid audios/songs outside phone library(cloud).
        //
        // Sometimes this filter won't work.
        cursor.addFilterPredicate(MPMediaPropertyPredicate.init(
            value: false,
            forProperty: MPMediaItemPropertyIsCloudItem,
            comparisonType: .equalTo
        ))
        
        // Using native sort from [IOS] you can only use the [Title], [Album] and
        // [Artist]. The others will be sorted 'manually' using [formatSongList] before
        // sending to Dart.
        cursor.groupingType = checkSongSortType(sortType: sortType)
        
        // Request permission status from the 'main' method.
        let hasPermission = SwiftOnAudioQueryPlugin().checkPermission()
        
        // We cannot 'query' without permission so, throw a PlatformException.
        // Only one 'channel' will be 'functional'. If is null, ignore, if not, send the error.
        if !hasPermission {
            // Call from 'EventChannel' (observer)
            self.sink?(
                FlutterError.init(
                    code: "403",
                    message: "The app doesn't have permission to read files.",
                    details: "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
                )
            )
            
            // Call from 'MethodChannel' (method)
            self.result?(
                FlutterError.init(
                    code: "403",
                    message: "The app doesn't have permission to read files.",
                    details: "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
                )
            )
            
            // 'Exit' the function
            return
        }
        
        // If [items] is null. Call early return with empty list.
        if cursor.items == nil {
            // Empty list.
            self.sink?([])
            self.result?([])
            
            // 'Exit' the function.
            return
        }
        
        // Query everything in background for a better performance.
        DispatchQueue.global(qos: .userInitiated).async {
            var listOfSongs: [[String: Any?]] = Array()
            
            // For each item(song) inside this "cursor", take one and "format"
            // into a [Map<String, dynamic>], all keys are based on [Android]
            // platforms so, if you change some key, will have to change the [Android] too.
            for song in cursor.items! {
                // If the song file don't has a assetURL, is a Cloud item.
                if !song.isCloudItem && song.assetURL != nil {
                    let songData = self.loadSongItem(song: song)
                    listOfSongs.append(songData)
                }
            }
            
            // Define the [toQuery] and [toRemove] filter.
            let toQuery = self.args["toQuery"] as! [Int: [String]]
            let toRemove = self.args["toRemove"] as! [Int: [String]]
            
            // 'Build' the filter.
            listOfSongs = listOfSongs.mediaFilter(
                mediaProjection: self.songProjection,
                toQuery: toQuery,
                toRemove: toRemove
            )
            
            // After finish the "query", go back to the "main" thread(You can only call flutter
            // inside the main thread).
            DispatchQueue.main.async {
                // Here we'll check the "custom" sort and define a order to the list.
                let finalList = self.formatSongList(allSongs: listOfSongs)
                
                // After loading the information, send the 'result'.
                self.sink?(finalList)
                self.result?(finalList)
            }
        }
    }
    
    private func loadSongItem(song: MPMediaItem) -> [String: Any?] {
        let fileExt = song.assetURL?.pathExtension ?? ""
        let sizeInBytes = song.value(forProperty: "fileSize") as? Int
        return [
            "_id": song.persistentID,
            "_data": song.assetURL?.absoluteString,
            "_uri": song.assetURL?.absoluteString,
            "_display_name": "\(song.artist ?? "") - \(song.title ?? "").\(fileExt)",
            "_display_name_wo_ext": "\(song.artist ?? "") - \(song.title ?? "")",
            "_size": sizeInBytes,
            "audio_id": nil,
            "album": song.albumTitle,
            "album_id": song.albumPersistentID,
            "artist": song.artist,
            "artist_id": song.artistPersistentID,
            "genre": song.genre,
            "genre_id": song.genrePersistentID,
            "bookmark": Int(song.bookmarkTime),
            "composer": song.composer,
            "date_added": Int(song.dateAdded.timeIntervalSince1970),
            "date_modified": 0,
            "duration": Int(song.playbackDuration * 1000),
            "title": song.title,
            "track": song.albumTrackNumber,
            "file_extension": fileExt,
        ]
    }
    
    private func formatSongList(allSongs: [[String: Any?]]) -> [[String: Any?]] {
        // Define a copy of all songs.
        var allSongsCopy = allSongs
        
        // Define all 'basic' filters.
        let order = args["orderType"] as? Int
        let sortType = args["sortType"] as? Int
        let ignoreCase = args["ignoreCase"] as! Bool
        
        // Sort the list 'manually'.
        switch sortType {
        case 3:
            allSongsCopy.sort { (val1, val2) in
                (val1["duration"] as! Double) > (val2["duration"] as! Double)
            }
        case 4:
            allSongsCopy.sort { (val1, val2) in
                (val1["date_added"] as! Int) > (val2["date_added"] as! Int)
            }
        case 5:
            allSongsCopy.sort { (val1, val2) in
                (val1["_size"] as! Int) > (val2["_size"] as! Int)
            }
        case 6:
            allSongsCopy.sort { (val1, val2) in
                ((val1["_display_name"] as! String).isCase(ignoreCase: ignoreCase)) > ((val2["_display_name"] as! String).isCase(ignoreCase: ignoreCase))
            }
        case 7:
            allSongsCopy.sort { (val1, val2) in
                (val1["track"] as! Int) > (val2["track"] as! Int)
            }
        default:
            break
        }
        
        // The order value is [1], reverse the list.
        return order == 1 ? allSongsCopy.reversed() : allSongsCopy
    }
}
